### PR TITLE
ExplicitStringVariableFixer: handle parsed array and object

### LINKED ISSUE #3332: explicit_string_variable fixes object property access incorrectly
When reporting an issue (bug) please provide the following information:

#### The PHP version you are using (`$ php -v`):
=> 7.2.0

#### PHP CS Fixer version you are using (`$ php-cs-fixer -V`):
=> 2.9.0-DEV

#### The command you use to run PHP CS Fixer:
=> vendor/bin/php-cs-fixer fix

#### The configuration file you are using, if any:
```php
<?php

$finder = PhpCsFixer\Finder::create()
    ->in(__DIR__)
    ->exclude('vendor');

return PhpCsFixer\Config::create()
    ->setRules(['explicit_string_variable' => true])
    ->setFinder($finder);
```

#### If applicable, please provide minimum samples of PHP code (as plain text, not screenshots):
 * before running PHP CS Fixer (no changes):
```php
<?php
echo "$a->b";
```

 * with unexpected changes applied when running PHP CS Fixer:
```php
<?php
echo "${a}->b";
```

 * with the changes you expected instead:
```php
<?php
echo "{$a->b}";
```

### PR DESCRIPTION
**Pull request by Slamdunk**
Closes #3332

### CODE REVIEW

File: src/Fixer/StringNotation/ExplicitStringVariableFixer.php [hunk 1]
```diff
@@ -59,17 +67,54 @@ protected function applyFix(\SplFileInfo $file, Tokens $tokens)
...
+            if (1 === count($variableTokens)) {
                 $tokens->overrideRange($index, $index, [
                     new Token([T_DOLLAR_OPEN_CURLY_BRACES, '${']),
                     new Token([T_STRING_VARNAME, substr($token->getContent(), 1)]),
                     new Token([CT::T_DOLLAR_CLOSE_CURLY_BRACES, '}']),
                 ]);
+            } else {
+                foreach ($variableTokens as $variablePartIndex => $variablePartToken) {
+                    if ($variablePartToken->isGivenKind(T_NUM_STRING)) {
+                        $tokens[$variablePartIndex] = new Token([T_LNUMBER, $variablePartToken->getContent()]);
```
Comment (Slamdunk) [inline src/Fixer/StringNotation/ExplicitStringVariableFixer.php]: OMG to this 😞 
I have to manually change the type of the token: is there a know map?
Maybe this is not the only one, although test cases cover all the allowed syntax cited in the manual:
https://secure.php.net/manual/en/language.types.string.php#language.types.string.parsing.simple

File: src/Fixer/StringNotation/ExplicitStringVariableFixer.php [hunk 2]
```diff
@@ -59,17 +67,54 @@ protected function applyFix(\SplFileInfo $file, Tokens $tokens)
...
+                    }
+                }
+
+                $tokens->insertAt($lastVariableTokenIndex + 1, new Token([CT::T_CURLY_CLOSE, '}']));
+                $tokens->insertAt($firstVariableTokenIndex, new Token([T_CURLY_OPEN, '{']));
```
Comment (keradus) [inline src/Fixer/StringNotation/ExplicitStringVariableFixer.php]: https://github.com/FriendsOfPHP/PHP-CS-Fixer/pull/3327#discussion_r157570814

Comment (Slamdunk) [reply src/Fixer/StringNotation/ExplicitStringVariableFixer.php]: I think you've reached the hundredth time you are talking about going backward. Maybe we should write an AutoReview?

BTW, fixed.

Comment (keradus) [reply src/Fixer/StringNotation/ExplicitStringVariableFixer.php]: it's not always needed to go backward. also, even if adding new tokens to collection, going backward is not only solution.

**Comment by keradus** [comment]
Thank you @Slamdunk.

**Pull request closed without merging**